src/materials/Material.js

```diff
@@ -19,6 +19,7 @@ function Material() {
 	this.side = FrontSide;
 	this.flatShading = false;
 	this.vertexColors = false;
+	this.vertexAlpha = false;
```

Comment:
@Mugen87 After #20135, do you think we still need to add `vertexAlpha`?

Reply:
This boolean controls whether the define `USE_VERTEX_ALPHA` is added to the shader or not. Depending on this define, the color attribute is `vec3` or `vec4`.

Reply:
But we can detect that automatically:

```js
geometry.attributes.color.itemSize === 4;
```

Reply:
Even if an attribute is present in the geometry, we do not generally force the user to use it. Hence, the material flags.

Are tangents an exception? We should be consistent in how we handle this...

Reply:
Well... We do not generally do it because we couldn't do it. Having these flags were a way to signal the user that materials couldn't always be reused. After #20135 we can.

Reply:
I agree that we should infer as many material flags as possible from geometry attributes/object flags. I’ll read #20135 related code more carefully later and see if ‘vertexAlpha’ can be inferred automatically.

We probably should start removing other flags as well? Of course not in this PR.

Reply:
> I agree that we should infer as many material flags as possible from geometry attributes/object flags. I’ll read #20135 related code more carefully later and see if ‘vertexAlpha’ can be inferred automatically.

Sound good 👍

I think this check should work:

```js
vertexAlpha: geometry.attributes.color && geometry.attributes.color.itemSize === 4,
```

> We probably should start removing other flags as well? Of course not in this PR.

Yep! I think we should be able to remove `vertexColors`, `vertexTangents` and `skinning` 💆‍♂️

Reply:
@mrdoob Although #20135 is merged, it does not mean materials properties like `vertexAlpha` are not necessary anymore. If we don't have these flags, we need code like in `setProgram()`:

https://github.com/mrdoob/three.js/blob/6cfe91706d1b5434f250d7be76cad29287926f58/src/renderers/WebGLRenderer.js#L1490-L1498

These checks are important since we only want to enter `getProgram()` when required.

Reply:
@WestLangley wrote:

>Even if an attribute is present in the geometry, we do not generally force the user to use it.

I do not think we should _infer_ material properties from the _geometry_.

Reply:
> I do not think we should infer material properties from the geometry.

@WestLangley Do you mind to explain the motivation?

For my motivation for inferring the flags, imagine the following case. A user may set the `color` attribute to be a 4-component Uint8Array, containing the R,G,B,A,R,G,B,A... values, but forget to set the `vertexAlpha` flag. WebGL will then silently discard the `A` values and the user may have to dig into `MeshStandardMaterial` and further `Material` documentation to see why. Inferring the `vertexAlpha` flag can avoid this case.

Reply:
@chubei We can't prevent user errors.

The following code would not be possible if material properties were inferred from the geometry:
https://github.com/mrdoob/three.js/blob/6deef54a1ece0fef56765885b19044db27469f04/editor/js/Viewport.js#L46-L48

Reply:
> We can't prevent user errors.

I agree that we can't prevent all of them, but we can prevent some of them, making the API easier to use.

`grid1.material.vertexColors = false; ` can be changed to `grid1.geometry.deleteAttribute('color');` for equivalent effect. Of course compatibility is a problem. I guess we have different opinions on whether it's worth it.

Anyway, this feature actually doesn't need either manual or automatic `vertexAlpha` flag. I just realized if you pass 3-component values to WebGL, it'll fill the last component with the default value for that attribute, which is default to 1.0. We can always use `vec4` in the shader.

Reply:
Sorry, but I do not vote to implement vertex color alpha like that since it's a breaking change. Custom shader code might break with this message:

> cannot convert from 'in highp 4-component vector of float' to 'out highp 3-component vector of float'

Reply:
Didn't realize it's a breaking change. Thanks for pointing that out @Mugen87. I'll reset to last reviewed state and continue investigating automatic inferring.

Reply:
@Do you mind resetting the PR to 07f1cb16a2e438a5b771415a442fb01a081d5b2b?

Reply:
> @chubei Do you mind resetting the PR to [07f1cb1](https://github.com/mrdoob/three.js/commit/07f1cb16a2e438a5b771415a442fb01a081d5b2b)?

Of course. Reset.

Reply:
> grid1.material.vertexColors = false; can be changed to grid1.geometry.deleteAttribute('color'); for equivalent effect.

This becomes problematic if you want to reuse the same geometry with different materials. According to your  approach, you would have to clone the geometry multiple times and configure it differently. This sounds quite way less efficient than just using different materials.

Reply:
> This becomes problematic if you want to render the same geometry with different effects. According to your approach, you would have to clone the geometry multiple times and configure it differently. This sounds less efficient than just using the same geometry but different materials.

Creating a new `geometry` without cloning underlying `TypedArray` should still be quite cheap? Not saying it’s better than material flags in terms of efficiency, just discussing possibilities.

Reply:
Even when performing a shallow clone the workflow still sounds strange to me.